--[[

CFamily

Language grammars for C-family and systems languages: C, C++, C#,
Java, Go, Rust, Zig, Kotlin, Swift, and Dart.

--]]

local DOUBLE_QUOTE = { open = '"', close = '"' }
local SINGLE_QUOTE = { open = "'", close = "'" }
local C_COMMENTS = { { "/*", "*/" } }
local C_LINE_COMMENTS = { "//" }

local C = {
    blockComments = C_COMMENTS,
    keywords = {
        "auto",
        "break",
        "case",
        "const",
        "continue",
        "default",
        "do",
        "else",
        "enum",
        "extern",
        "for",
        "goto",
        "if",
        "inline",
        "register",
        "restrict",
        "return",
        "sizeof",
        "static",
        "struct",
        "switch",
        "typedef",
        "union",
        "volatile",
        "while",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "char",
        "double",
        "float",
        "int",
        "long",
        "short",
        "signed",
        "unsigned",
        "void",
        "bool",
        "size_t",
        "int8_t",
        "int16_t",
        "int32_t",
        "int64_t",
        "uint8_t",
        "uint16_t",
        "uint32_t",
        "uint64_t",
        "NULL",
    },
}

local Cpp = {
    blockComments = C_COMMENTS,
    keywords = {
        "alignas",
        "alignof",
        "and",
        "and_eq",
        "asm",
        "auto",
        "bitand",
        "bitor",
        "break",
        "case",
        "catch",
        "class",
        "co_await",
        "co_return",
        "co_yield",
        "compl",
        "concept",
        "const",
        "consteval",
        "constexpr",
        "constinit",
        "continue",
        "decltype",
        "default",
        "delete",
        "do",
        "dynamic_cast",
        "else",
        "enum",
        "explicit",
        "export",
        "extern",
        "false",
        "for",
        "friend",
        "goto",
        "if",
        "inline",
        "mutable",
        "namespace",
        "new",
        "noexcept",
        "not",
        "not_eq",
        "nullptr",
        "operator",
        "or",
        "or_eq",
        "private",
        "protected",
        "public",
        "register",
        "reinterpret_cast",
        "requires",
        "return",
        "sizeof",
        "static",
        "static_assert",
        "static_cast",
        "struct",
        "switch",
        "template",
        "this",
        "throw",
        "true",
        "try",
        "typedef",
        "typeid",
        "typename",
        "union",
        "using",
        "virtual",
        "volatile",
        "while",
        "xor",
        "xor_eq",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = 'R"(', close = ')"', multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "char",
        "char8_t",
        "char16_t",
        "char32_t",
        "double",
        "float",
        "int",
        "long",
        "short",
        "signed",
        "size_t",
        "string",
        "unsigned",
        "void",
        "wchar_t",
        "auto",
        "nullptr_t",
        "vector",
        "map",
        "set",
        "pair",
        "tuple",
        "shared_ptr",
        "unique_ptr",
    },
}

local CSharp = {
    blockComments = C_COMMENTS,
    keywords = {
        "abstract",
        "as",
        "async",
        "await",
        "base",
        "break",
        "case",
        "catch",
        "checked",
        "class",
        "const",
        "continue",
        "default",
        "delegate",
        "do",
        "else",
        "enum",
        "event",
        "explicit",
        "extern",
        "false",
        "finally",
        "fixed",
        "for",
        "foreach",
        "goto",
        "if",
        "implicit",
        "in",
        "interface",
        "internal",
        "is",
        "lock",
        "namespace",
        "new",
        "null",
        "operator",
        "out",
        "override",
        "params",
        "private",
        "protected",
        "public",
        "readonly",
        "record",
        "ref",
        "return",
        "sealed",
        "sizeof",
        "stackalloc",
        "static",
        "struct",
        "switch",
        "this",
        "throw",
        "true",
        "try",
        "typeof",
        "unchecked",
        "unsafe",
        "using",
        "var",
        "virtual",
        "volatile",
        "while",
        "yield",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = '@"', close = '"', escape = '""' },
        { open = '$"', close = '"' },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "byte",
        "char",
        "decimal",
        "double",
        "dynamic",
        "float",
        "int",
        "long",
        "nint",
        "nuint",
        "object",
        "sbyte",
        "short",
        "string",
        "uint",
        "ulong",
        "ushort",
        "void",
    },
}

local Java = {
    blockComments = C_COMMENTS,
    keywords = {
        "abstract",
        "assert",
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "default",
        "do",
        "else",
        "enum",
        "extends",
        "false",
        "final",
        "finally",
        "for",
        "goto",
        "if",
        "implements",
        "import",
        "instanceof",
        "interface",
        "native",
        "new",
        "null",
        "package",
        "private",
        "protected",
        "public",
        "return",
        "sealed",
        "static",
        "strictfp",
        "super",
        "switch",
        "synchronized",
        "this",
        "throw",
        "throws",
        "transient",
        "true",
        "try",
        "var",
        "void",
        "volatile",
        "while",
        "yield",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = '"""', close = '"""', multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "boolean",
        "byte",
        "char",
        "double",
        "float",
        "int",
        "long",
        "short",
        "String",
        "Integer",
        "Boolean",
        "Double",
        "Float",
        "Long",
        "Short",
        "Byte",
        "Character",
        "Object",
        "List",
        "Map",
        "Set",
        "Optional",
    },
    uppercaseTypes = true,
}

local Go = {
    blockComments = C_COMMENTS,
    builtins = {
        "append",
        "cap",
        "close",
        "complex",
        "copy",
        "delete",
        "imag",
        "len",
        "make",
        "new",
        "panic",
        "print",
        "println",
        "real",
        "recover",
    },
    keywords = {
        "break",
        "case",
        "chan",
        "const",
        "continue",
        "default",
        "defer",
        "else",
        "fallthrough",
        "for",
        "func",
        "go",
        "goto",
        "if",
        "import",
        "interface",
        "map",
        "package",
        "range",
        "return",
        "select",
        "struct",
        "switch",
        "type",
        "var",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = "`", close = "`", multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "byte",
        "complex64",
        "complex128",
        "error",
        "float32",
        "float64",
        "int",
        "int8",
        "int16",
        "int32",
        "int64",
        "rune",
        "string",
        "uint",
        "uint8",
        "uint16",
        "uint32",
        "uint64",
        "uintptr",
        "any",
    },
    uppercaseTypes = true,
}

local Rust = {
    blockComments = C_COMMENTS,
    builtins = {
        "assert",
        "assert_eq",
        "assert_ne",
        "cfg",
        "column",
        "compile_error",
        "concat",
        "dbg",
        "debug_assert",
        "debug_assert_eq",
        "debug_assert_ne",
        "env",
        "eprint",
        "eprintln",
        "file",
        "format",
        "include",
        "include_bytes",
        "include_str",
        "line",
        "macro_rules",
        "matches",
        "module_path",
        "option_env",
        "panic",
        "print",
        "println",
        "stringify",
        "todo",
        "unimplemented",
        "unreachable",
        "vec",
        "write",
        "writeln",
    },
    keywords = {
        "as",
        "async",
        "await",
        "break",
        "const",
        "continue",
        "crate",
        "dyn",
        "else",
        "enum",
        "extern",
        "false",
        "fn",
        "for",
        "if",
        "impl",
        "in",
        "let",
        "loop",
        "match",
        "mod",
        "move",
        "mut",
        "pub",
        "ref",
        "return",
        "self",
        "Self",
        "static",
        "struct",
        "super",
        "trait",
        "true",
        "type",
        "unsafe",
        "use",
        "where",
        "while",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = 'r#"', close = '"#', multiline = true },
        { open = 'r"', close = '"' },
        { open = "b'", close = "'" },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "char",
        "f32",
        "f64",
        "i8",
        "i16",
        "i32",
        "i64",
        "i128",
        "isize",
        "str",
        "u8",
        "u16",
        "u32",
        "u64",
        "u128",
        "usize",
        "Box",
        "Option",
        "Result",
        "String",
        "Vec",
        "HashMap",
        "HashSet",
        "Rc",
        "Arc",
        "Cell",
        "RefCell",
    },
    uppercaseTypes = true,
}

local Zig = {
    blockComments = {},
    builtins = {
        "@addWithOverflow",
        "@alignCast",
        "@alignOf",
        "@as",
        "@atomicLoad",
        "@atomicStore",
        "@bitCast",
        "@boolToInt",
        "@breakpoint",
        "@call",
        "@cDefine",
        "@ceil",
        "@cImport",
        "@cInclude",
        "@clz",
        "@compileError",
        "@compileLog",
        "@ctz",
        "@divExact",
        "@divFloor",
        "@divTrunc",
        "@embedFile",
        "@enumToInt",
        "@errSetCast",
        "@errorName",
        "@errorReturnTrace",
        "@export",
        "@fence",
        "@field",
        "@fieldParentPtr",
        "@floatCast",
        "@floatToInt",
        "@frame",
        "@frameAddress",
        "@import",
        "@intCast",
        "@intToEnum",
        "@intToFloat",
        "@intToPtr",
        "@max",
        "@memcpy",
        "@memset",
        "@min",
        "@mod",
        "@mulAdd",
        "@mulWithOverflow",
    },
    keywords = {
        "addrspace",
        "align",
        "allowzero",
        "and",
        "anyframe",
        "anytype",
        "asm",
        "async",
        "await",
        "break",
        "callconv",
        "catch",
        "comptime",
        "const",
        "continue",
        "defer",
        "else",
        "enum",
        "errdefer",
        "error",
        "export",
        "extern",
        "false",
        "fn",
        "for",
        "if",
        "inline",
        "linksection",
        "noalias",
        "nosuspend",
        "null",
        "opaque",
        "or",
        "orelse",
        "packed",
        "pub",
        "resume",
        "return",
        "struct",
        "suspend",
        "switch",
        "test",
        "threadlocal",
        "true",
        "try",
        "undefined",
        "union",
        "unreachable",
        "var",
        "volatile",
        "while",
    },
    lineComments = { "//" },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "comptime_float",
        "comptime_int",
        "f16",
        "f32",
        "f64",
        "f80",
        "f128",
        "i8",
        "i16",
        "i32",
        "i64",
        "i128",
        "isize",
        "noreturn",
        "type",
        "u8",
        "u16",
        "u32",
        "u64",
        "u128",
        "usize",
        "void",
    },
}

local Kotlin = {
    blockComments = C_COMMENTS,
    keywords = {
        "abstract",
        "actual",
        "annotation",
        "as",
        "break",
        "by",
        "catch",
        "class",
        "companion",
        "const",
        "constructor",
        "continue",
        "crossinline",
        "data",
        "do",
        "else",
        "enum",
        "expect",
        "external",
        "false",
        "final",
        "finally",
        "for",
        "fun",
        "if",
        "import",
        "in",
        "infix",
        "init",
        "inline",
        "inner",
        "interface",
        "internal",
        "is",
        "it",
        "lateinit",
        "noinline",
        "null",
        "object",
        "open",
        "operator",
        "out",
        "override",
        "package",
        "private",
        "protected",
        "public",
        "reified",
        "return",
        "sealed",
        "super",
        "suspend",
        "this",
        "throw",
        "true",
        "try",
        "typealias",
        "val",
        "var",
        "vararg",
        "when",
        "where",
        "while",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = '"""', close = '"""', multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "Any",
        "Boolean",
        "Byte",
        "Char",
        "Double",
        "Float",
        "Int",
        "Long",
        "Nothing",
        "Short",
        "String",
        "Unit",
    },
    uppercaseTypes = true,
}

local Swift = {
    blockComments = C_COMMENTS,
    keywords = {
        "Any",
        "Protocol",
        "Self",
        "Type",
        "as",
        "associatedtype",
        "async",
        "await",
        "break",
        "case",
        "catch",
        "class",
        "continue",
        "default",
        "defer",
        "deinit",
        "do",
        "else",
        "enum",
        "extension",
        "fallthrough",
        "false",
        "fileprivate",
        "for",
        "func",
        "guard",
        "if",
        "import",
        "in",
        "indirect",
        "infix",
        "init",
        "inout",
        "internal",
        "is",
        "lazy",
        "let",
        "mutating",
        "nil",
        "nonmutating",
        "open",
        "operator",
        "optional",
        "override",
        "postfix",
        "prefix",
        "private",
        "protocol",
        "public",
        "repeat",
        "required",
        "rethrows",
        "return",
        "self",
        "some",
        "static",
        "struct",
        "subscript",
        "super",
        "switch",
        "throw",
        "throws",
        "true",
        "try",
        "typealias",
        "var",
        "weak",
        "where",
        "while",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = '"""', close = '"""', multiline = true },
        DOUBLE_QUOTE,
    },
    types = {
        "Array",
        "Bool",
        "Character",
        "Dictionary",
        "Double",
        "Float",
        "Int",
        "Optional",
        "Set",
        "String",
        "UInt",
        "Void",
    },
    uppercaseTypes = true,
}

local Dart = {
    blockComments = C_COMMENTS,
    keywords = {
        "abstract",
        "as",
        "assert",
        "async",
        "await",
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "covariant",
        "default",
        "deferred",
        "do",
        "dynamic",
        "else",
        "enum",
        "export",
        "extends",
        "extension",
        "external",
        "factory",
        "false",
        "final",
        "finally",
        "for",
        "get",
        "hide",
        "if",
        "implements",
        "import",
        "in",
        "interface",
        "is",
        "late",
        "library",
        "mixin",
        "new",
        "null",
        "on",
        "operator",
        "part",
        "required",
        "rethrow",
        "return",
        "sealed",
        "set",
        "show",
        "static",
        "super",
        "switch",
        "sync",
        "this",
        "throw",
        "true",
        "try",
        "typedef",
        "var",
        "void",
        "when",
        "while",
        "with",
        "yield",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        { open = '"""', close = '"""', multiline = true },
        { open = "'''", close = "'''", multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "double",
        "dynamic",
        "int",
        "num",
        "void",
        "String",
        "List",
        "Map",
        "Set",
        "Future",
        "Stream",
        "Object",
    },
}

return {
    { aliases = { "c" }, grammar = C },
    { aliases = { "cpp", "c++", "cc", "cxx" }, grammar = Cpp },
    { aliases = { "csharp", "cs", "c#" }, grammar = CSharp },
    { aliases = { "dart" }, grammar = Dart },
    { aliases = { "go", "golang" }, grammar = Go },
    { aliases = { "java" }, grammar = Java },
    { aliases = { "kotlin", "kt" }, grammar = Kotlin },
    { aliases = { "rust", "rs" }, grammar = Rust },
    { aliases = { "swift" }, grammar = Swift },
    { aliases = { "zig" }, grammar = Zig },
}
