--[[

Scripting

Language grammars for scripting languages: Lua, Luau, Python, Ruby,
PHP, and Elixir.

--]]

local SharedConstants = require("./SharedConstants")

local DOUBLE_QUOTE = SharedConstants.DOUBLE_QUOTE
local SINGLE_QUOTE = SharedConstants.SINGLE_QUOTE

local Lua = {
    blockComments = { { "--[[", "]]" } },
    keywords = {
        "and",
        "break",
        "do",
        "else",
        "elseif",
        "end",
        "false",
        "for",
        "function",
        "if",
        "in",
        "local",
        "nil",
        "not",
        "or",
        "repeat",
        "return",
        "then",
        "true",
        "until",
        "while",
    },
    lineComments = { "--" },
    strings = {
        { open = "[[", close = "]]", multiline = true },
        { open = "[=[", close = "]=]", multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
}

local Luau = {
    blockComments = { { "--[[", "]]" } },
    keywords = {
        "and",
        "break",
        "continue",
        "do",
        "else",
        "elseif",
        "end",
        "export",
        "false",
        "for",
        "function",
        "if",
        "in",
        "local",
        "nil",
        "not",
        "or",
        "repeat",
        "return",
        "then",
        "true",
        "type",
        "until",
        "while",
    },
    lineComments = { "--" },
    strings = {
        { open = "[[", close = "]]", multiline = true },
        { open = "[=[", close = "]=]", multiline = true },
        { open = "`", close = "`" },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "any",
        "boolean",
        "buffer",
        "never",
        "nil",
        "number",
        "string",
        "thread",
        "unknown",
    },
}

local Python = {
    builtins = {
        "abs",
        "all",
        "any",
        "bin",
        "bool",
        "bytes",
        "chr",
        "dict",
        "dir",
        "enumerate",
        "eval",
        "exec",
        "filter",
        "float",
        "format",
        "getattr",
        "globals",
        "hasattr",
        "hash",
        "hex",
        "id",
        "input",
        "int",
        "isinstance",
        "issubclass",
        "iter",
        "len",
        "list",
        "locals",
        "map",
        "max",
        "min",
        "next",
        "object",
        "oct",
        "open",
        "ord",
        "pow",
        "print",
        "property",
        "range",
        "repr",
        "reversed",
        "round",
        "set",
        "setattr",
        "slice",
        "sorted",
        "staticmethod",
        "str",
        "sum",
        "super",
        "tuple",
        "type",
        "vars",
        "zip",
    },
    keywords = {
        "False",
        "None",
        "True",
        "and",
        "as",
        "assert",
        "async",
        "await",
        "break",
        "class",
        "continue",
        "def",
        "del",
        "elif",
        "else",
        "except",
        "finally",
        "for",
        "from",
        "global",
        "if",
        "import",
        "in",
        "is",
        "lambda",
        "nonlocal",
        "not",
        "or",
        "pass",
        "raise",
        "return",
        "try",
        "while",
        "with",
        "yield",
    },
    lineComments = { "#" },
    strings = {
        { open = '"""', close = '"""', multiline = true },
        { open = "'''", close = "'''", multiline = true },
        { open = 'f"', close = '"' },
        { open = "f'", close = "'" },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
}

local Ruby = {
    builtins = {
        "attr_accessor",
        "attr_reader",
        "attr_writer",
        "include",
        "extend",
        "prepend",
        "puts",
        "print",
        "p",
        "raise",
        "require",
        "require_relative",
    },
    keywords = {
        "BEGIN",
        "END",
        "alias",
        "and",
        "begin",
        "break",
        "case",
        "class",
        "def",
        "defined?",
        "do",
        "else",
        "elsif",
        "end",
        "ensure",
        "false",
        "for",
        "if",
        "in",
        "module",
        "next",
        "nil",
        "not",
        "or",
        "redo",
        "rescue",
        "retry",
        "return",
        "self",
        "super",
        "then",
        "true",
        "undef",
        "unless",
        "until",
        "when",
        "while",
        "yield",
    },
    lineComments = { "#" },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
}

local PHP = {
    blockComments = { { "/*", "*/" } },
    builtins = {
        "array",
        "count",
        "echo",
        "empty",
        "explode",
        "implode",
        "isset",
        "print_r",
        "strlen",
        "strpos",
        "substr",
        "unset",
        "var_dump",
    },
    keywords = {
        "abstract",
        "and",
        "as",
        "break",
        "case",
        "catch",
        "class",
        "clone",
        "const",
        "continue",
        "declare",
        "default",
        "do",
        "else",
        "elseif",
        "enddeclare",
        "endfor",
        "endforeach",
        "endif",
        "endswitch",
        "endwhile",
        "extends",
        "false",
        "final",
        "finally",
        "fn",
        "for",
        "foreach",
        "function",
        "global",
        "goto",
        "if",
        "implements",
        "include",
        "instanceof",
        "interface",
        "match",
        "namespace",
        "new",
        "null",
        "or",
        "private",
        "protected",
        "public",
        "readonly",
        "require",
        "return",
        "static",
        "switch",
        "throw",
        "trait",
        "true",
        "try",
        "use",
        "var",
        "while",
        "xor",
        "yield",
    },
    lineComments = { "//", "#" },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "bool",
        "float",
        "int",
        "mixed",
        "never",
        "null",
        "object",
        "self",
        "string",
        "void",
    },
}

local Elixir = {
    builtins = {
        "hd",
        "inspect",
        "is_atom",
        "is_binary",
        "is_boolean",
        "is_float",
        "is_function",
        "is_integer",
        "is_list",
        "is_map",
        "is_nil",
        "is_number",
        "is_pid",
        "is_tuple",
        "length",
        "puts",
        "raise",
        "tl",
    },
    keywords = {
        "after",
        "and",
        "case",
        "catch",
        "cond",
        "def",
        "defimpl",
        "defmacro",
        "defmodule",
        "defp",
        "defprotocol",
        "defstruct",
        "do",
        "else",
        "end",
        "false",
        "fn",
        "for",
        "if",
        "import",
        "in",
        "nil",
        "not",
        "or",
        "raise",
        "receive",
        "require",
        "rescue",
        "true",
        "try",
        "unless",
        "use",
        "when",
        "with",
    },
    lineComments = { "#" },
    strings = {
        { open = '"""', close = '"""', multiline = true },
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
}

return {
    { aliases = { "elixir", "ex" }, grammar = Elixir },
    { aliases = { "lua" }, grammar = Lua },
    { aliases = { "luau" }, grammar = Luau },
    { aliases = { "php" }, grammar = PHP },
    { aliases = { "python", "py" }, grammar = Python },
    { aliases = { "ruby", "rb" }, grammar = Ruby },
}
