--[[

WebJS

Language grammars for web languages: JavaScript, TypeScript, JSX,
TSX, CSS, HTML, and XML.

--]]

local SharedConstants = require("./SharedConstants")

local C_COMMENTS = SharedConstants.C_COMMENTS
local C_LINE_COMMENTS = SharedConstants.C_LINE_COMMENTS
local DOUBLE_QUOTE = SharedConstants.DOUBLE_QUOTE
local SINGLE_QUOTE = SharedConstants.SINGLE_QUOTE
local TEMPLATE_LITERAL = { open = "`", close = "`", multiline = true }

local JavaScript = {
    blockComments = C_COMMENTS,
    builtins = {
        "Array",
        "Boolean",
        "Date",
        "Error",
        "Function",
        "Infinity",
        "JSON",
        "Map",
        "Math",
        "NaN",
        "Number",
        "Object",
        "Promise",
        "Proxy",
        "Reflect",
        "RegExp",
        "Set",
        "String",
        "Symbol",
        "WeakMap",
        "WeakSet",
        "clearInterval",
        "clearTimeout",
        "console",
        "decodeURI",
        "decodeURIComponent",
        "document",
        "encodeURI",
        "encodeURIComponent",
        "eval",
        "fetch",
        "globalThis",
        "isFinite",
        "isNaN",
        "parseInt",
        "parseFloat",
        "require",
        "setInterval",
        "setTimeout",
        "undefined",
        "window",
    },
    keywords = {
        "async",
        "await",
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "debugger",
        "default",
        "delete",
        "do",
        "else",
        "export",
        "extends",
        "false",
        "finally",
        "for",
        "from",
        "function",
        "if",
        "import",
        "in",
        "instanceof",
        "let",
        "new",
        "null",
        "of",
        "return",
        "static",
        "super",
        "switch",
        "this",
        "throw",
        "true",
        "try",
        "typeof",
        "var",
        "void",
        "while",
        "with",
        "yield",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        TEMPLATE_LITERAL,
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
}

local TypeScript = {
    blockComments = C_COMMENTS,
    builtins = {
        "Array",
        "Boolean",
        "Date",
        "Error",
        "Function",
        "Infinity",
        "JSON",
        "Map",
        "Math",
        "NaN",
        "Number",
        "Object",
        "Promise",
        "Proxy",
        "Reflect",
        "RegExp",
        "Set",
        "String",
        "Symbol",
        "WeakMap",
        "WeakSet",
        "clearInterval",
        "clearTimeout",
        "console",
        "decodeURI",
        "decodeURIComponent",
        "document",
        "encodeURI",
        "encodeURIComponent",
        "eval",
        "fetch",
        "globalThis",
        "isFinite",
        "isNaN",
        "parseInt",
        "parseFloat",
        "require",
        "setInterval",
        "setTimeout",
        "undefined",
        "window",
    },
    keywords = {
        "abstract",
        "as",
        "async",
        "await",
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "debugger",
        "declare",
        "default",
        "delete",
        "do",
        "else",
        "enum",
        "export",
        "extends",
        "false",
        "finally",
        "for",
        "from",
        "function",
        "if",
        "implements",
        "import",
        "in",
        "infer",
        "instanceof",
        "interface",
        "is",
        "keyof",
        "let",
        "module",
        "namespace",
        "never",
        "new",
        "null",
        "of",
        "override",
        "package",
        "private",
        "protected",
        "public",
        "readonly",
        "return",
        "satisfies",
        "static",
        "super",
        "switch",
        "this",
        "throw",
        "true",
        "try",
        "type",
        "typeof",
        "undefined",
        "var",
        "void",
        "while",
        "with",
        "yield",
    },
    lineComments = C_LINE_COMMENTS,
    strings = {
        TEMPLATE_LITERAL,
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    types = {
        "any",
        "bigint",
        "boolean",
        "never",
        "null",
        "number",
        "object",
        "string",
        "symbol",
        "undefined",
        "unknown",
        "void",
        "Partial",
        "Required",
        "Readonly",
        "Record",
        "Pick",
        "Omit",
        "Exclude",
        "Extract",
        "NonNullable",
        "ReturnType",
        "Parameters",
        "Awaited",
    },
}

local CSS = {
    blockComments = C_COMMENTS,
    keywords = {
        "!important",
        "@charset",
        "@font-face",
        "@import",
        "@keyframes",
        "@media",
        "@page",
        "@supports",
        "and",
        "from",
        "not",
        "only",
        "or",
        "to",
    },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    detectFunctionCalls = false,
}

local HTML = {
    blockComments = { { "<!--", "-->" } },
    keywords = {
        "a",
        "abbr",
        "article",
        "aside",
        "audio",
        "body",
        "br",
        "button",
        "canvas",
        "code",
        "details",
        "div",
        "em",
        "footer",
        "form",
        "h1",
        "h2",
        "h3",
        "h4",
        "h5",
        "h6",
        "head",
        "header",
        "hr",
        "html",
        "iframe",
        "img",
        "input",
        "label",
        "li",
        "link",
        "main",
        "meta",
        "nav",
        "ol",
        "p",
        "pre",
        "script",
        "section",
        "select",
        "source",
        "span",
        "strong",
        "style",
        "summary",
        "table",
        "td",
        "template",
        "textarea",
        "th",
        "title",
        "tr",
        "ul",
        "video",
    },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    detectFunctionCalls = false,
    htmlTags = true,
}

local XML = {
    blockComments = { { "<!--", "-->" } },
    strings = {
        DOUBLE_QUOTE,
        SINGLE_QUOTE,
    },
    detectFunctionCalls = false,
    htmlTags = true,
}

return {
    { aliases = { "css" }, grammar = CSS },
    { aliases = { "html", "htm" }, grammar = HTML },
    { aliases = { "javascript", "js" }, grammar = JavaScript },
    { aliases = { "jsx" }, grammar = JavaScript },
    { aliases = { "typescript", "ts" }, grammar = TypeScript },
    { aliases = { "tsx" }, grammar = TypeScript },
    { aliases = { "xml", "svg" }, grammar = XML },
}
